### ISSUE TITLE
[CSA] False negative: constant row-bound violations in multidimensional arrays are missed

### ISSUE THREAD

**Issue by rikki322** [llvm/llvm-project#214223]
Hi, I found that `security.ArrayBound` misses constant second-dimension subscripts that leave a row but remain inside the enclosing array's flat storage.

## Program

```c
int matrix[10][10];

void overflow_row(void) {
  matrix[2][10] = 1;
}

void underflow_row(void) {
  matrix[5][-1] = 1;
}
```

Both subscripts violate the bound of the selected `int[10]` row. CSA emits no `security.ArrayBound` result for either access.

## CSA reproduction

```sh
clang --analyze \
  -Xclang -analyzer-checker=security.ArrayBound \
  multidimensional_array_row_bounds.c
```

## Comparison with related programs

A direct flat out-of-bounds access is reported:

```c
int matrix[10][10];

void direct(void) {
  ((int *)matrix)[100] = 1; // warning: out of bound access
}
```

The missed program uses a constant subscript that leaves the selected inner row while still landing inside the enclosing array's flat storage:

```c
int matrix[10][10];

void overflow_row(void) {
  matrix[2][10] = 1; // no warning
}

void underflow_row(void) {
  matrix[5][-1] = 1; // no warning
}
```

## Runtime confirmation

Compiling this reproducer with `-fsanitize=bounds` reports an out-of-bounds index for both the overflow and underflow variants.

## Expected result

`security.ArrayBound` should retain the bounds of each nested array dimension instead of treating these accesses as valid offsets into flat outer storage.

## Verification

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by NagyDonat** [comment]
I do have plans for strengthening the checks of `security.ArrayBound` in the near future. 

However, note that ArrayBound is torn between requests like this that request more strict checks and requests like https://github.com/llvm/llvm-project/pull/214140 that ask for further relaxation of the bounds checking. I'll try to aim for a compromise, but I'm not sure what should be the default.
